streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java

```diff
@@ -77,31 +80,51 @@ public StateStore getStateStore(final String name) {
 
     @SuppressWarnings("unchecked")
     @Override
+    public <K, V> void forward(final K key, final V value) {
+        forward(key, value, SEND_TO_ALL);
+    }
+
+    @SuppressWarnings({"unchecked", "deprecation"})
+    @Override
     public <K, V> void forward(final K key, final V value, final int childIndex) {
-        final ProcessorNode previousNode = currentNode();
-        final ProcessorNode child = (ProcessorNode<K, V>) currentNode().children().get(childIndex);
-        setCurrentNode(child);
-        try {
-            child.process(key, value);
-        } finally {
-            setCurrentNode(previousNode);
-        }
+        forward(key, value, To.child(((List<ProcessorNode>) currentNode().children()).get(childIndex).name()));
     }
 
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({"unchecked", "deprecation"})
     @Override
     public <K, V> void forward(final K key, final V value, final String childName) {
-        for (ProcessorNode child : (List<ProcessorNode<K, V>>) currentNode().children()) {
-            if (child.name().equals(childName)) {
-                ProcessorNode previousNode = currentNode();
+        forward(key, value, To.child(childName));
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <K, V> void forward(final K key, final V value, final To to) {
+        toAccessor.update(to);
+        if (toAccessor.hasTimestamp()) {
+            recordContext.setTimestamp(toAccessor.timestamp());
+        }
+        final ProcessorNode previousNode = currentNode();
+        try {
+            final List<ProcessorNode<K, V>> children = (List<ProcessorNode<K, V>>) currentNode().children();
+            final String sendTo = toAccessor.child();
+            if (sendTo != null) {
+                final ProcessorNode child = currentNode().getChild(sendTo);
                 setCurrentNode(child);
-                try {
+                child.process(key, value);
+            } else {
+                if (children.size() == 1) {
```

Comment:
why do we need this? isn't it just a special case of the else below?

Reply:
It's a special case that allows us to avoid creating the iterator object.